pyvrp/cpp/search/Route.h

```diff
@@ -235,29 +235,29 @@ Distance Route::distBetween(size_t start, size_t end) const
 {
     assert(start <= end && end <= nodes.size() + 1);
 
-    auto const startDist = start == 0 ? 0 : nodes[start - 1]->cumulatedDistance;
-    auto const endDist = end == nodes.size() + 1
-                             ? endDepot.cumulatedDistance
-                             : nodes[end - 1]->cumulatedDistance;
+    if (end == 0)
+        return 0;
 
-    assert(startDist <= endDist);
+    auto const startDist = start == 0 ? 0 : cumDist[start - 1];
+    auto const endDist = end == nodes.size() + 1 ? distance_ : cumDist[end - 1];
 
+    assert(startDist <= endDist);
     return endDist - startDist;
 }
 
 Load Route::loadBetween(size_t start, size_t end) const
 {
     assert(start <= end && end <= nodes.size() + 1);
 
+    if (end == 0)
+        return 0;
+
     auto const *startNode = start == 0 ? &startDepot : nodes[start - 1];
     auto const atStart = data.client(startNode->client).demand;
-    auto const startLoad = startNode->cumulatedLoad;
-    auto const endLoad = end == nodes.size() + 1
-                             ? endDepot.cumulatedLoad
-                             : nodes[end - 1]->cumulatedLoad;
+    auto const startLoad = start == 0 ? 0 : cumLoad[start - 1];
+    auto const endLoad = end == nodes.size() + 1 ? load_ : cumLoad[end - 1];
```

Comment:
There are a lot of branches here that are hard to predict right. I'd like to simplify these to something like
```cpp
  return cum[end] - cum[start];
```
or similar. That's much easier to understand, and should give us a nice performance boost.

To get there, the `pyvrp::search::Route` object should probably start tracking the start/end depots in these lists explicitly. I think that's OK if we word it very carefully. The upside is that then the distinction between `position` and `index` falls away as well.

Reply:
This sounds great! While considering this it may be worth to keep in mind #51 (proposal to define `end` to be exclusive following usual conventions, e.g. numpy slicing) as well?